drivers/i2c/i2c_ll_stm32.c

```diff
@@ -226,6 +226,133 @@ static int i2c_stm32_transfer(const struct device *dev, struct i2c_msg *msg,
 	return ret;
 }
 
+#ifdef CONFIG_I2C_CALLBACK
+
+static void i2c_stm32_async_done(const struct device *dev, int result)
+{
+	struct i2c_stm32_data *data = dev->data;
+	i2c_callback_t cb = data->cb;
+	void *userdata = data->userdata;
+
+	data->cb = NULL;
+	data->userdata = NULL;
+
+	k_sem_give(&data->bus_mutex);
+
+	/* Callback may wish to start another transfer */
+	cb(dev, result, userdata);
+}
+
+static int i2c_stm32_async_iter(const struct device *dev, struct i2c_msg msg,
+				uint8_t *next_msg_flags, uint16_t slave)
+{
+	int ret;
+
+	if (msg.len > 0) {
+		i2c_stm32_async_done(dev, 0);
+		return 0;
+	}
+
+	ret = stm32_i2c_transaction(dev, msg, next_msg_flags, slave);
+	if (ret < 0) {
+		i2c_stm32_async_done(dev, ret);
+	}
+
+	return ret;
+}
+
+static int i2c_stm32_transfer_cb(const struct device *dev, struct i2c_msg *msg, uint8_t num_msgs,
```

Comment:
This should return -EBUSY if the peripheral is already in use. It should not take a semaphore locking anything as it may be called from an ISR